sparkle/output/configuration_output.py

```diff
@@ -0,0 +1,256 @@
+            instance_set_train: Instance set used for training
+            instance_set_test: Instance set used for testing
+            penalty_multiplier: penalty multiplier that is applied to the par performance
+                [To be Removed]
+            output: Path to the output directory
+        """
+        self.solver = solver
+        self.configurator = configurator
+        self.instance_set_train = instance_set_train
+        self.instance_set_test = instance_set_test
+        self.penalty_multiplier = penalty_multiplier
+        self.directory = path
+
+        if output is None:
+            output = path / "Analysis" / "configuration.json"
+        self.output = output
+
+        # TODO: Fix this spaghetti code to find the scenario file
+        solver_dir_name = path.name
+        scenario_dir = path / "outdir_train_configuration" / \
+            f"{solver_dir_name}_scenario" / "state-run0"
+        if scenario_dir.is_dir():
+            scenario_file = scenario_dir / "scenario.txt"
+            if not scenario_file.is_file():
+                print("Can't find scenario file")
+                return
+        else:
+            print("Can't find scenario file")
+            return
+
+        scenario_dir = scenario_dir.parent
+        # Sets scenario on configurator object
+        self.configurator.scenario = \
+            ConfigurationScenario.from_file(scenario_file, self.solver,
+                                            self.instance_set_train,
+                                            scenario_dir)
+        self.configurator.scenario._set_paths(self.configurator.output_path)
+
+        # Retrieve all configurations
+        self.configurations = self.get_configurations(path)
+
+        # Retrieves validation results and best configuration
+        self.training, self.best_config =\
+            self.get_validation_data(self.instance_set_train)
+
+        # Retrieve test validation results if they exist
+        if self.instance_set_test is not None:
+            self.test, _ = self.get_validation_data(self.instance_set_test)
+
+    def parse_string_to_dict(self: ConfigurationOutput, input_string: str) -> dict:
+        """Convert a configuration string to a dictionary."""
+        result_dict = {}
+        input_string = input_string.replace("'", "")
+        input_string = input_string.replace(",", "")
+        input_string = input_string.replace(":", "")
+        input_string = input_string.replace("{", "")
+        input_string = input_string.replace("}", "")
+        tokens = input_string.split()
+
+        for i in range(0, len(tokens), 2):
+            key = tokens[i]
+            value = tokens[i + 1]
+            result_dict[key] = value
+
+        return result_dict
+
+    def get_configurations(self: ConfigurationOutput, path: Path) -> list[dict]:
+        """Read all configurations and transform them to dictionaries."""
+        config_path = path / "validation" / "configurations.csv"
```

Comment:
Hardcoded path, this should be passed directly to the method (Complete path to the .csv)

Reply:
I understand that we don't want to use hardcoded paths, but how should I do it without just using a hard-coded path in the function which calls get_configurations?
Somewhere the path needs to be defined. I did move it outside now

Reply:
Try to do it at the highest level you can, e.g. where this method is being called I think in this case. That way we have all that at least grouped together and keeps the methods quite flexible. But aren't you able to do this using the validator class?